nm: Fix incorrect device-reapply call

cathay4t opened the pull request:
When creating OVS port for OVS system interface, nmstate use the same name
for both OVS system interface connection and OVS port connection, the
`org.freedesktop.NetworkManager.GetDeviceByIpIface` will return unexpected
`NmDevice` when we try to do `device-reapply` call.

To fix this issue, all `NmApi::connection_reapply()` call should search all
existing `NmDevice` by both interface name and type.

To achieve that this patch unified the NmDevice.iface_type (integer) with
NmConnectionSetting.iface_type (string) into `enum NmifaceType`.

Integration test case included.

Resolves: https://issues.redhat.com/browse/RHEL-50556

cathay4t committed:
nm: Change unknown device state log to debug

Signed-off-by: Gris Ge <fge@redhat.com>

liangwen12year commented on tests/integration/ovs_test.py:
```diff
@@ -2201,6 +2201,28 @@ def remove_ovn_state_present(state):
+@ip_monitor_assert_stable_link_up(BOND1)
+def test_ovs_system_iface_link_stable(ovs_bridge1_with_bond_as_system_iface):
+    desired_state = ovs_bridge1_with_bond_as_system_iface
+    libnmstate.apply(desired_state)
```
Maybe we can add `assertlib.assert_state_match(desired_state)` in the end.

cathay4t replied:
Added

cathay4t replied:
`assert_state_match` does not works well when ovs bridge and ovs interface are using the same name.

And checking state matches is not the key point of this test case.

So I removed the state match check line.

ihuguet commented on rust/src/lib/nm/nm_dbus/query_apply/device.rs:
```diff
@@ -16,15 +17,17 @@ const NM_DEVICE_TYPE_WIFI: u32 = 2;
 const NM_DEVICE_TYPE_BT: u32 = 5;
 const NM_DEVICE_TYPE_OLPC_MESH: u32 = 6;
-const NM_DEVICE_TYPE_WIMAX: u32 = 7;
+// WIMIX is not supported by NM
```
typo: WIMIX -> WIMAX

ihuguet commented on rust/src/lib/nm/nm_dbus/query_apply/iface_type.rs:
I would move the content of this file to nm_dbus/connection/iface_type.rs. Otherwise, when we modify something related to iface_types in the future, it's easy to miss that this file exists. It's easier to modify/extend something if everything is together in one place. Of course, there are good reasons to split related things into different files, but I think that in this case it's better to keep them together.

ihuguet commented on rust/src/lib/nm/query_apply/device.rs:
```diff
@@ -1,43 +1,39 @@
     InterfaceType,
 };
 
+fn nm_iface_type_to_nmstate(nm_iface_type: &NmIfaceType) -> InterfaceType {
+    match nm_iface_type {
+        NmIfaceType::Bridge => InterfaceType::LinuxBridge,
+        NmIfaceType::Bond => InterfaceType::Bond,
+        NmIfaceType::Ethernet => InterfaceType::Ethernet,
+        NmIfaceType::OvsBridge => InterfaceType::OvsBridge,
+        NmIfaceType::OvsIface => InterfaceType::OvsInterface,
+        NmIfaceType::Vlan => InterfaceType::Vlan,
+        NmIfaceType::Vxlan => InterfaceType::Vxlan,
+        NmIfaceType::Dummy => InterfaceType::Dummy,
+        NmIfaceType::Macvlan => InterfaceType::MacVlan,
+        NmIfaceType::Vrf => InterfaceType::Vrf,
+        // We unify the Veth to into InterfaceType::Ethernet to simplify
+        // work on using veth as plain ethernet interface
+        NmIfaceType::Veth => InterfaceType::Ethernet,
+        NmIfaceType::Infiniband => InterfaceType::InfiniBand,
+        NmIfaceType::Loopback => InterfaceType::Loopback,
+        NmIfaceType::Macsec => InterfaceType::MacSec,
+        NmIfaceType::Hsr => InterfaceType::Hsr,
+        _ => InterfaceType::Other(nm_iface_type.to_string()),
+    }
+}
```
Same, this function would be better placed at nm_dbus/connection/iface_type.rs

cathay4t replied:
The `nm_dbus` cannot depend on `nmstate::InterfaceType`. The `nm_dbus` folder was supposed to be standalone and could be released as a standalone crate in the future.

ihuguet commented on rust/src/lib/nm/settings/connection.rs:
```diff
@@ -349,38 +314,26 @@ pub(crate) fn iface_to_nm_connections(
 
 pub(crate) fn iface_type_to_nm(
```
Same.

ihuguet reviewed (commented):
Apart from those nits, LGTM

cathay4t committed:
nm: Fix incorrect device-reapply call

When creating OVS port for OVS system interface, nmstate use the same name
for both OVS system interface connection and OVS port connection,
the `org.freedesktop.NetworkManager.GetDeviceByIpIface` will return
unexpected `NmDevice` when we try to do `device-reapply` call.

To fix this issue, all `NmApi::connection_reapply()` call should search
all existing `NmDevice` by both interface name and type.

To achieve that this patch unified the NmDevice.iface_type (integer) with
NmConnectionSetting.iface_type (string) into `enum NmifaceType`.

Integration test case included.

Resolves: https://issues.redhat.com/browse/RHEL-50556

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
I have waited long enough and this PR is also tested by our support engineer in lab. Force merging.
